README.md
<h1 id="agavedb.keyval">AgaveDB</h1>

[![Build Status](https://travis-ci.org/TACC/agavedb.svg?branch=master)](https://travis-ci.org/TACC/agavedb)

AgaveDB is multiuser-aware key/value store built using the Agave metadata web service API. 

The library interface is modeled on [pickledb](https://pythonhosted.org/pickleDB/), which, in turn is
inspired by Redis. Eventually, it will support Agave-based permissions and sharing. If you need a more
document-oriented solution, you can utilize the underlying Agave `metadata` service.

<h2 id="example">Using AgaveDB</h2>

An implicit assumption is that you have an existing client set up using AgavePy or the TACC Cloud CLI. 

Installation is simple because [AgaveDB is on PyPi](https://pypi.python.org/pypi/agavedb): `pip install agavedb`. Then, start coding. AgaveDB doesn't have a CLI at present. 

```python
>>> from agavedb import AgaveKeyValStore, Agave
>>> ag = Agave.restore()
>>> db = AgaveKeyValStore(ag)
>>> db.set('key1', 'value1')
'key1'
>>> db.set('key2', 'value2')
'key2'
>>> db.set('key3', 'value4')
True
>>> db.get('key1')
u'value1'
>>> db.set('key1', 'value3')
'key1'
>>> db.get('key1')
u'value3'
>>> db.getall()
[u'key1', u'key2', u'key3']
>>> db.rem('key1')
True
>>> db.getall()
[u'key2', u'key3']
>>> db.set(db.genid(), 'hello world')
'MZgY69k1ZMd8'
>>> db.get('MZgY69k1ZMd8')
u'hello world'
>>> db.deldb()
True
>>> db.getall()
[]
```

<h2 id="keys-values">About Keys and Values</h2>

<h3 id="keys">Keys</h3>

AgaveDB keys are strings. You have a good degree of flexibility in naming them but there are rules:

* Minimum key length is 4 characters. Really short keys aren't necessary. You will probably thank yourself for writing `user:1001:bugreports` instead of `u1001bg` when it comes time to debug or maintain our code. The added space and memory requirement is trivial. 
* Maximum key length is 512 characters, but such long keys aren't a great idea because they will consume memory and will be slow to retrieve. Consider using a hash (such as SHA1) to represent a larger value. 
* Schemas are good. Examples include `object-type:id` and `user:1001`. Dots, dashes, and underscores are often used as separators.
* Whitespace is not allowed in keynames nor are the characters `/` and `#`

<h3 id="values">Values</h3>

AgaveDB values can be strings (or Unicode) or numeric values. Formally, this is validated by checking whether a value an instance of Python's `basestring` or `numbers.Number` primitives. Support for lists, dicts, and tuples will be added in future releases. The maximum size for an AgaveDB value is currently 4096 bytes. 

<h3 id="apidocs">API Documentation</h3>

Automatically generated API docs [are here](docs/api.md)

<h3 id="apidocs">Tests</h3>

Tests are implemented using `pytest`. Documentation is in the [tests folder](agavedb/tests/README.md)

<h3 id="source">Source Install</h3>

In addition to using PyPyi, you can install AgaveDB from this source repository.

**From a Github Checkout**

```shell
$ git clone https://github.com/TACC/agavedb
$ cd agavedb
$ git checkout <branch_name>
$ python setup.py install
```

**Via a GitHub URL**

```shell
$ pip install git+https://github.com/TACC/agavedb.git@<branch>
```



CHANGELOG.md
# AgaveDB

All significant changes will be recorded in this document

## 0.4.0 - Refactored unicode and testing

## 0.3.0 - ACLs

* Added
    * Support for user-specific ACL via setacl, getacls, remacl. Needed to support TACC Reactors 'alias' functions.

* Changed
    * Updated documentation

* Removed
    * Nothing

## 0.2.2 - Versioned keys

* Added
    * Internal support for arbitrary or versioned key prefixes
    * Unit testing

* Changed
    * Nothing

* Removed
    * Nothing

## 0.1.7 - Bug fix

* Added
    * Nothing

* Changed
    * Now safe to pass urlencoded strings as key names

* Removed
    * Nothing

## 0.1.5 - Python3 compatible

* Added
    * Python 3.3 and 3.6 compatibility

* Changed
    * Moved to Sphinx + ReadTheDocs for documentation

* Removed
    * Nothing


## 0.1.1 - Standalone Release

* Added
    * Packaged for PyPi

* Changed
    * Nothing

* Removed
    * Nothing

## 0.1.0 - First release

* Added

* Changed

* Removed


agavedb/tests/testdata.py
__author__ = 'mwvaughn'

import json
import os


HERE = os.path.dirname(os.path.abspath(__file__))


class TestData(object):

    def __init__(self, credentials):
        self.local_data = credentials
        self.dat = self.file_to_json('testdata.json')

    def file_to_json(self, filename):
        return json.load(open(os.path.join(HERE, filename)))

    def data(self, key=None):
        if key is None:
            return self.dat
        else:
            return self.dat.get(key, None)


agavedb/tests/agavefixtures.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Usage: See bundled README.md

from future.standard_library import install_aliases
install_aliases()
from urllib.parse import urlencode, quote
from past.builtins import basestring

import json
import os
import pytest
import random
import uuid

from agavepy.agave import Agave, AgaveError

HERE = os.path.dirname(os.path.abspath(__file__))
ENV_PREFIXES = ('AGAVE', '_AGAVE', 'TACC', '_TACC')


@pytest.fixture(scope='session')
def credentials():
    '''
    Load credentials for testing session

    Order: user credential store, test file, env
    '''
    credentials = {}
    # credential store
    # credential store
    if os.environ.get('AGAVE_CACHE_DIR', None) is not None:
        ag_cred_store = os.path.join(
            os.environ.get('AGAVE_CACHE_DIR'), 'current')
    else:
        ag_cred_store = os.path.expanduser('~/.agave/current')

    if os.path.exists(ag_cred_store):
        tempcred = json.load(open(ag_cred_store, 'r'))
        credentials['apiserver'] = tempcred.get('baseurl', None)
        credentials['username'] = tempcred.get('username', None)
        credentials['password'] = tempcred.get('password', None)
        credentials['apikey'] = tempcred.get('apikey', None)
        credentials['apisecret'] = tempcred.get('apisecret', None)
        credentials['token'] = tempcred.get('access_token', None)
        credentials['refresh_token'] = tempcred.get('refresh_token', None)
        credentials['verify_certs'] = tempcred.get('verify', None)
        credentials['client_name'] = tempcred.get('client_name', None)
        credentials['tenantid'] = tempcred.get('tenantid', None)
    # test file
    credentials_file = os.environ.get('creds', 'test_credentials.json')
    if os.path.exists(credentials_file):
        credentials = json.load(open(
            os.path.join(HERE, credentials_file), 'r'))
    # environment
    for env in ('apikey', 'apisecret', 'username', 'password',
                'apiserver', 'verify_certs', 'refresh_token',
                'token', 'client_name'):
        for vpx in ENV_PREFIXES:
            varname = '_'.join([vpx, env.upper()])
            if os.environ.get(varname, None) is not None:
                credentials[env] = os.environ.get(varname)

    return credentials


@pytest.fixture(scope='session')
def agave(credentials):
    ag = Agave(username=credentials.get('username'),
               password=credentials.get('password', None),
               client_name=credentials.get('client_name', None),
               api_server=credentials.get('apiserver'),
               api_key=credentials.get('apikey'),
               api_secret=credentials.get('apisecret'),
               token=credentials.get('token', None),
               refresh_token=credentials.get('refresh_token', None),
               verify=True)
    return ag


scripts/checkrepo.sh
#!/bin/bash

require_clean_work_tree () {
    # Update the index
    git update-index -q --ignore-submodules --refresh
    err=0

    # Disallow unstaged changes in the working tree
    if ! git diff-files --quiet --ignore-submodules --
    then
        echo >&2 "cannot $1: you have unstaged changes."
        git diff-files --name-status -r --ignore-submodules -- >&2
        err=1
    fi

    # Disallow uncommitted changes in the index
    if ! git diff-index --cached --quiet HEAD --ignore-submodules --
    then
        echo >&2 "cannot $1: your index contains uncommitted changes."
        git diff-index --cached --name-status -r --ignore-submodules HEAD -- >&2
        err=1
    fi

    if [ $err = 1 ]
    then
        echo >&2 "Please commit or stash them."
        exit 1
    fi
}

require_clean_work_tree


agavedb/keyval.py
"""
AgaveDB

Multiuser-aware key/value store built atop the Agave metadata API.

The library interface is modeled on pickledb, which is inspired by Redis, but
with the addition of simple access control list via the Agave permisisons
model. If you need document-oriented solution, you should use the actual
Agave metadata service rather than AgaveDB.

Usage:
```python
from agavedb import AgaveKeyValStore, Agave

db = AgaveKeyValStore(Agave.restore())
```
"""
from __future__ import print_function
from __future__ import absolute_import
from future.standard_library import install_aliases
install_aliases()

__version__ = '0.4.6'

from past.builtins import basestring
from agavepy.agave import Agave, AgaveError

import base64
import re
import json
import logging
import time
import sys
import os
from . import uniqueid

_SEP = '/'
_PREFIX = '_agkvs_v2'
_TTL = 86400
_MAX_VAL_BYTES = 4096
_MIN_KEY_BYTES = 4
_MAX_KEY_BYTES = 512
_RE_KEY_NAMES = re.compile('^[\S]+$', re.UNICODE)
VALID_PEMS = ['read', 'write', 'execute']
VALID_ROLE_USERNAMES = ['world', 'public']
# TODO: Implement list and dict methods
# DONE: Implement permissions management


class AgaveKeyValStore(object):

    """An AgaveKeyValStore instance. Requires an active Agave client"""

    def __init__(self, agaveClient, prefix=_PREFIX, loglevel='INFO'):
        """
        Initialize a AgaveKeyValStore object

        Positional parameters:
        agaveClient - an initialaized Agave object

        Keyword parameters:
        prefix - str - Optional override for key prefix
        loglevel - Set the logging level. DEBUG is good for diagnosing issues.

        Returns:
        - AgaveKeyValStore
        """
        self.client = agaveClient
        self.default_ttl = _TTL
        FORMAT = "%(asctime)s [%(levelname)s] - %(message)s"
        DATEFORMAT = "%Y-%m-%dT%H:%M:%SZ"
        self.logging = logging.getLogger('AgaveKeyValStore')
        self.logging.setLevel(loglevel)
        stderrLogger = logging.StreamHandler()
        stderrLogger.setFormatter(
            logging.Formatter(FORMAT, datefmt=DATEFORMAT))
        self.logging.addHandler(stderrLogger)
        self.prefix = prefix
        self.version = __version__

    def set(self, key, value):
        '''Set the string or numeric value of a key'''
        try:
            return self._set(key, value)
        except Exception as e:
            self.logging.error("set({}, {}): {}".format(key, value, e))
            return None

    def setacl(self, key, acl):
        """
        Set an ACL on a given key

        Positional parameters:
        key - str - Key to manage permissions on
        acl - dict - Valid permissions object

        {'read': bool, 'write': bool}

        Returns:
        Boolean True on success, and Exception + False on failure
        """
        try:
            self.validate_acl(acl)
            return self._setacl(key, acl)
        except Exception as e:
            self.logging.debug(
                "Failed to set ACl({}, {}): {}".format(key, acl, e))
            return False

    def remacl(self, key, user):
        """
        Remove an ACL on a key for a given user

        Positional parameters:
        key - str - Key to manage permissions on
        user - str - Username whose ACL should be dropped

        Returns:
        Boolean True on success, and Exception + False on failure
        """
        acl = {'username': user,
               'permission': {'read': None, 'write': None, 'execute': None}}
        try:
            return self._setacl(key, acl)
        except Exception as e:
            self.logging.debug(
                "Failed to remove ACL({}, {}): {}".format(key, user, e))
            return False

    def getacls(self, key, user=None):
        """
        Get the ACLs for a given key

        Keywork parameters:
        user - str - Return ACL only for this username
        """
        try:
            return self._getacls(key, user)
        except Exception as e:
            self.logging.debug("Failed to getacls: {}".format(e))
            return []

    def get(self, key):
        '''Get the value of a key'''
        try:
            return self._get(key)['value']
        except Exception as e:
            self.logging.debug("Failed to get: {}".format(e))
            return None

    def getall(self, sorted=True):
        '''Return a list of all keys user owns or has access to'''
        try:
            return self._getall(sorted=sorted, namespace=False)
        except Exception as e:
            self.logging.debug("Failed to getall: {}".format(e))
            return []

    def rem(self, key):
        '''Delete a key (assuming it is owned by the user)'''
        try:
            return self._rem(key)
        except Exception as e:
            self.logging.debug("Failed to rem: {}".format(e))
            return False

    def deldb(self):
        '''Delete all user-owned keys'''
        try:
            return self._remall()
        except Exception as e:
            self.logging.debug("Failed to deldb: {}".format(e))
            return False

    def _stringify(self, value):
        if isinstance(value, bytes):
            return value.decode('utf-8')
        else:
            return value

    def _namespace(self, keyname, permissive=True):
        """Namespaces a key

        e.g.) keyname => _agavedb/keyname#username

        Returns: string-like object
        """
        if self._key_is_valid(keyname):
            _keyname = base64.urlsafe_b64encode(keyname.encode()).decode()
            _keyname = self.prefix + _SEP + _keyname + '#' + self._username()
            #return str(_keyname)
            # make it stringy
            return self._stringify(_keyname)
        else:
            if permissive is False:
                raise ValueError("Invalid key name: {}".format(keyname))
            return None

    def _rev_namespace(self, fullkeyname,
                       removeusername=True, permissive=True):
        """Reverse namespacing of a key's internal representation.

        e.g.) _agavedb/keyname#username => keyname

        Returns: string-like object
        """
        assert isinstance(fullkeyname, basestring), \
            "key type must be string or unicode (type: {})".format(
                self._type(fullkeyname))

        _prefix = self.prefix + _SEP
        keyname = fullkeyname
        if keyname.startswith(_prefix):
            keyname = keyname[len(_prefix):]

        # if removeusername:
        _suffix = '#' + self._username()
        if keyname.endswith(_suffix):
            keyname = keyname[:(-1 * len(_suffix))]

        # print(keyname)
        if not isinstance(keyname, bytes):
            keyname = keyname.encode()

        keyname_tmp = ''
        keyname_tmp = base64.urlsafe_b64decode(keyname)
        keyname = keyname_tmp

        # Python2/3 compatible coercion to a "stringy" key name
        if isinstance(keyname, bytes):
            if not removeusername:
                keyname = keyname + _suffix.encode()
            try:
                return keyname.decode('utf-8')
            except Exception as e:
                if permissive:
                    return None
                else:
                    raise Exception(e)
        else:
            if not removeusername:
                keyname = keyname + _suffix
            try:
                return str(keyname)
            except Exception as e:
                if permissive:
                    return None
                else:
                    raise Exception(e)

    def _slugify(self, value, allow_unicode=False):
        """
        Convert a string to a conservatively URL-safe version

        Converts to ASCII if 'allow_unicode' is False. Converts
        whitespace to hyphens. Removes characters that aren't
        alphanumerics, underscores, or hyphens. Converts to
        lowercase. Strips leading and trailing whitespace.
        """
        import unicodedata
        value = str(value)
        if allow_unicode:
            value = unicodedata.normalize('NFKC', value)
        else:
            value = unicodedata.normalize(
                'NFKD', value).encode('ascii', 'ignore').decode('ascii')
        value = re.sub(r'[^\w\s-]', '', value).strip().lower()
        value = re.sub(r'[-\s]+', '-', value)
         # stringify
        return self._stringify(value)
        # if isinstance(value, bytes):
        #     return value.decode('utf-8')
        # else:
        #     return value

    def _type(self, obj):
        '''Return the type name of a Python object'''
        return type(obj).__name__

    def _username(self):
        '''Return the current Agave API username'''
        try:
            return self._stringify(self.__get_api_username())
        except Exception as e:
            raise Exception("Unable to establish username: {}".format(e))
            return None

    def _get(self, key):
        '''Get value by key name.'''
        shares = False
        key_name = self._namespace(key)
        username = self._username()
        # An Agave metadata object, not yet the value
        if shares:
            _regex = "^{}/{}#".format(self.prefix, key)
            query = json.dumps({'name': {'$regex': _regex, '$options': 'i'}})
        else:
            query = json.dumps({'name': key_name})

        key_objs = self.client.meta.listMetadata(q=query)
        key_objs_owner = []
        key_objs_other = []
        for key_obj in key_objs:
            if key_obj['owner'] == username:
                key_objs_owner.append(key_obj)
            else:
                key_objs_other.append(key_obj)
        key_objs_merged = key_objs_owner + key_objs_other
        if len(key_objs_merged) > 0:
            return self._stringify(key_objs_merged[0])
        else:
            raise KeyError("No such key: {}".format(key))

    def _set(self, key, value):
        '''Update/write value to a key'''
        key_name = self._namespace(key)
        key_uuid = None

        if not self._value_is_valid(value):
            raise ValueError(
                "Key type for {} not valid (type: {})".format(
                    key, self._type(value)))
            return None

        key_uuid_obj = {}
        try:
            key_uuid_obj = self._get(key)
            key_uuid = key_uuid_obj['uuid']
        except KeyError:
            self.logging.debug("Key {} doesn't yet exist".format(key))
            pass

        current_time = int(time.time())
        if '_created' in key_uuid_obj:
            created_t = key_uuid_obj['_created']
            expires_t = current_time + _TTL
        else:
            created_t = current_time
            expires_t = created_t + _TTL

        try:
            value = str(value)
        except Exception as e:
            self.logging.debug(
                "Couldn't coerce {} to unicode: {}".format(value, e))
            return None

        try:
            value = self._mongo_stringify(value)
        except Exception as e:
            self.logging.debug(
                "Couldn't stringify {}: {}".format(value, e))
            return None

        # our metadata record with timestamps
        meta = json.dumps({'name': key_name,
                           'value': value,
                           '_created': created_t,
                           '_expires': expires_t,
                           '_ttl': _TTL}, separators=(',', ':'))

        if key_uuid is None:
            # Create
            try:
                self.client.meta.addMetadata(body=meta)
            except Exception as e:
                self.logging.debug("Error creating key {}: {}".format(key, e))
                return None
        else:
            # Update
            try:
                self.client.meta.updateMetadata(uuid=key_uuid, body=meta)
            except Exception as e:
                self.logging.debug("Error updating key {}: {}".format(key, e))
                return None

        return self._stringify(key)

    def _setacl(self, key, acl):
        '''Add or update an ACL to a key'''
        key_uuid = None
        key_uuid_obj = {}
        try:
            key_uuid_obj = self._get(key)
            key_uuid = key_uuid_obj['uuid']
        except KeyError:
            self.logging.debug("Key {} not found".format(key))
            raise KeyError("Key {} not found".format(key))

        pem = self.to_text_pem(acl)
        meta = json.dumps(pem, separators=(',', ':'))
        try:
            self.client.meta.updateMetadataPermissions(
                uuid=key_uuid, body=meta)
        except Exception as e:
                self.logging.debug(
                    "Error setting ACL for {}: {}".format(key, e))
                return False

        return True

    def _getacls(self, key, user=None):
        '''List ACLs on a given key'''
        key_uuid = None
        key_uuid_obj = {}
        acls = []

        try:
            key_uuid_obj = self._get(key)
            key_uuid = key_uuid_obj['uuid']
        except KeyError:
            self.logging.debug("Key {} not found".format(key))
            raise KeyError("Key {} not found".format(key))
        try:
            resp = self.client.meta.listMetadataPermissions(uuid=key_uuid)
            for acl in resp:
                formatted_acl = {'username': acl.get('username'),
                                 'permission': acl.get('permission')}
                if user is None:
                    acls.append(formatted_acl)
                else:
                    if user == acl.get('username'):
                        acls.append(formatted_acl)
                    # show the user is inheriting world acl
                    elif 'world' == acl.get('username'):
                        acls.append(formatted_acl)
        except Exception as e:
            self.logging.debug(
                "Failed getting ACLs for for {}: {}".format(key, e))

        return acls

    def _getall(self, sorted=True, namespace=False, uuids=False):
        '''Fetch and return all keys visible to the user'''
        all_keys = []
        _regex = "^{}/*".format(self.prefix)
        query = json.dumps({'name': {'$regex': _regex, '$options': 'i'}})
        # collection of Agave metadata objects
        key_objs = self.client.meta.listMetadata(q=query)
        for key_obj in key_objs:
            if uuids:
                tmp_uuid = key_obj['uuid']
                if isinstance(tmp_uuid, bytes):
                    tmp_uuid = tmp_uuid.decode('utf-8')
                all_keys.append(tmp_uuid)
            elif namespace:
                temp_key = self._rev_namespace(
                    key_obj['name'], removeusername=namespace)
                all_keys.append(temp_key)
            else:
                temp_key = self._rev_namespace(key_obj['name'])
                all_keys.append(temp_key)

        if sorted:
            all_keys.sort()

        safened_keys = []
        for k in all_keys:
            safened_keys.append(self._stringify(k))

        return safened_keys

    def _rem(self, key):
        '''Delete a key from a user's namespace'''
        key_uuid = None
        try:
            key_uuid = self._get(key)
            key_uuid = key_uuid['uuid']
        except KeyError:
            raise KeyError("No such key: {}".format(key))
            return False
        except Exception as e:
            self.logging.debug("Error validating key {}: {}".format(key, e))
            return False

        try:
            self._rem_by_uuid(key_uuid)
        except Exception as e:
            self.logging.debug("Error deleting key {}".format(key))
            return False
        return True

    def _rem_by_uuid(self, key_uuid):
        '''Delete key by its UUID'''
        try:
            self.client.meta.deleteMetadata(uuid=key_uuid)
        except Exception:
            self.logging.debug(
                "Error deleting key with UUID {}".format(key_uuid))
            return False
        return True

    def _remall(self):
        '''Remove all the user's keys'''
        try:
            key_list = self._getall(uuids=True)
            for key_uuid in key_list:
                self._rem_by_uuid(key_uuid)
        except Exception:
            self.logging.debug("Error deleting all keys")
            return False
        return True

    def _value_is_valid(self, value):
        '''Value must be a string. Others may be supported later.'''
        assert isinstance(value, basestring), \
            "value must be str-like (type: {})".format(
                self._type(value))
        assert len(value) <= _MAX_VAL_BYTES, \
            "value must be <= {} bytes (length: {})".format(len(value))
        return True

    def _key_is_valid(self, key):
        '''Enforce key naming restrictions'''

        # type
        assert isinstance(key, basestring), \
            "key type must be str-like type (type: {})".format(
                self._type(key))

        # character set
        assert _RE_KEY_NAMES.match(key), \
            "key may only contain non-whitespace characters"

        # now that the key value is base64 encoded before namespacing, this
        # is no longer required
        # assert _SEP not in key, "key may not contain '{}'".format(_SEP)
        # assert '#' not in key, "key may not contain #"

        # length
        assert len(key) <= _MAX_KEY_BYTES, \
            "key must be <= {} characters (length: {})".format(
                _MAX_KEY_BYTES, len(key))
        assert len(key) >= _MIN_KEY_BYTES, \
            "key must be >= {} characters (length: {})".format(
                _MIN_KEY_BYTES, len(key))

        return True

    def _mongo_stringify(self, value):
        '''Coerce a value to a string type before sending to MongoDB'''
        return '"' + value + '"'

    @classmethod
    def create_key_name(cls):
        '''Create a unique, human-friendly key name'''
        return uniqueid.get_id()

    @classmethod
    def to_text_pem(cls, acl):
        pem = {"username": acl.get('username')}
        permission = acl.get('permission')
        perm_str = u'NONE'
        r, w, x = permission.get('read', False), \
            permission.get('write', False), \
            permission.get('execute', False)

        if r:
            perm_str = u'READ'
        if w:
            perm_str = u'READ_WRITE'
        if x:
            perm_str = u'READ_EXECUTE'
        if r and w:
            perm_str = u'READ_WRITE'
        if r and x:
            perm_str = u'READ_EXECUTE'
        if r and w and x:
            perm_str = u'ALL'

        pem['permission'] = perm_str
        return pem

    @classmethod
    def from_text_pem(cls, pem):
        acl = {"username": pem.get('username')}
        permission = pem.get('permission').upper()
        perm_dict = {'read': False, 'write': False, 'execute': False}

        if u'ALL' in permission:
            perm_dict = {'read': True, 'write': True, 'execute': True}
        elif u'NONE' in permission:
            perm_dict = {'read': False, 'write': False, 'execute': False}
        else:
            if u'READ' in permission:
                perm_dict['read'] = True
            if u'WRITE' in permission:
                perm_dict['write'] = True
                perm_dict['read'] = True
            if u'EXECUTE' in permission:
                perm_dict['execute'] = True
                perm_dict['read'] = True
        acl['permission'] = perm_dict
        return acl

    @classmethod
    def validate_acl(cls, acl, permissive=False):
        """
        Validate an ACL object as a dict

        Failure raises Exception unless permissive is True
        * Does not validate that username exists
        """
        err = 'Invalid ACL: {}'
        try:
            assert isinstance(acl, dict), "Not a dict"
            assert 'username' in acl and 'permission' in acl, \
                "Both username and permission are required"
            assert isinstance(acl['permission'], dict), \
                "Permission must be a dict"
            assert isinstance(acl['username'], basestring), \
                "Username must be string or unicode"
            assert set(acl['permission'].keys()) == set(VALID_PEMS) or \
                set(acl['permission'].keys()) <= set(VALID_PEMS), \
                "Valid permission types are {} not {}".format(
                    VALID_PEMS, list(acl['permission'].keys()))
            for p in acl['permission']:
                assert isinstance(acl['permission'][p], bool), \
                    "Only Boolean values allowed for permission values"
            return True
        except Exception as exc:
            if permissive is True:
                return False
            else:
                raise AgaveError(err.format(exc))

    def __get_api_username(self):
        '''Determine username'''
        if os.environ.get('_abaco_username'):
            return os.environ.get('_abaco_username')
        elif self.client.username is not None:
            return self.client.username
        else:
            self.logging.debug("No username could be determined")
            return None

    def __get_api_token(self):
        '''Determine API access_token'''
        if os.environ.get('_abaco_access_token'):
            return os.environ.get('_abaco_access_token')
        elif self.client.token.token_info.get('access_token') is not None:
            return self.client.token.token_info.get('access_token')
        else:
            self.logging.debug("Failed to retrieve API access_token")
            return None

    def __get_api_server(self):
        '''Determine API server'''
        if os.environ.get('_abaco_api_server'):
            return os.environ.get('_abaco_api_server')
        elif self.client.token.api_server is not None:
            return self.client.token.api_server
        else:
            self.logging.debug("Returning hard-coded value for API server")
            return 'https://api.sd2e.org'


def main():
    ag = Agave.restore()
    kvs = AgaveKeyValStore(ag)

    print(kvs.getall())

if __name__ == '__main__':
    main()
